src/reflect/scala/reflect/internal/Variances.scala

```diff
@@ -85,12 +89,14 @@ trait Variances {
           else v
 
         @tailrec
-        def loop(sym: Symbol, v: Variance): Variance = (
-          if (sym == tvar.owner || v.isBivariant) v
+        def loop(sym: Symbol, v: Variance): Variance =
+          if (v.isBivariant) v
+          else if (sym == tvar.owner) if (lowerBoundStack.contains(sym)) v.flip else v
```

Comment:
Could the `if (lowerBoundStack.contains(sym)) v.flip` logic be moved into `nextVariance`/`shouldFlip`, and be handled before we get to this step of the recursion? We have a list of cases of when `shouldFlip` is true; would be great if you could add `lowerBoundStack.contains(sym.owner)` in there, if that makes sense.

In any case, please add some more comments around the usage of `lowerBoundStack`.

Reply:
That was my attempt initially. But the recursion starts at `loop` and goes into `nextVariance` later (in the next branch). In some cases we don't even call `nextVariance` at all. Another option would be to merge `loop` and `nextVariance`.

Agree about the comments, I was thinking about adding more to the `validateDefinition` method.

Reply:
Ok, thanks for explaining. Would be good to also render those findings as comments, because it's likely someone else will have the same thought again at some point, if we already have evidence for N=2 :-)

Reply:
@adriaanm I need a little help with something. I don't see `apply` being reentrant in any way (which confused me because the original version was saving and restoring the previous `base`).
Doesn't that mean we can replace the stack with a single variable?

Reply:
Trying this now and also added documentation